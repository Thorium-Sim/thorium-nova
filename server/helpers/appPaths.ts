import path from "path";
import fs from "fs";
import os from "os";

if (!fs.existsSync(`${os.homedir()}/Documents`)) {
  fs.mkdirSync(`${os.homedir()}/Documents`);
}

export let thoriumPath = `${os.homedir()}/Documents/thorium`;

if (process.env.THORIUM_PATH) {
  const testPath = String(process.env.THORIUM_PATH).replace("~", os.homedir());
  try {
    fs.mkdirSync(testPath, {recursive: true});
    thoriumPath = testPath;
  } catch {
    // Do nothing.
  }
}

export let assetDir = path.resolve("./assets");

if (process.env.NODE_ENV === "production") {
  assetDir = `${thoriumPath}/assets`;
}

export let appStoreDir = "./data/";

if (process.env.NODE_ENV === "production") {
  appStoreDir = thoriumPath + "/";
}

const appStoreName =
  process.env.NODE_ENV === "production"
    ? "snapshot.json"
    : process.env.NODE_ENV === "test"
    ? "snapshot-test.json"
    : "snapshot-dev.json";

export const appStorePath = `${appStoreDir}${appStoreName}`;
