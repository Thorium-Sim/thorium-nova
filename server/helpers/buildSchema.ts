import {buildSchema as buildTypeSchema} from "type-graphql";
import {pubsub} from "./pubsub";
import {resolvers} from "../schema";

export default function buildSchema() {
  return buildTypeSchema({
    resolvers,
    pubSub: pubsub,
    validate: false,
  });
}
