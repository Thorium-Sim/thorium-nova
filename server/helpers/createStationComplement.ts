import {StationComplementComponent} from "server/components/stationComplement";
import {Card} from "server/schema/card";
import Station from "server/schema/station";

import Entity from "./ecs/entity";

const Pilot = new Station({
  name: "Pilot",
});
const stationNames = [
  ["Pilot"],
  ["Navigation", "Tactical"],
  ["Navigation", "Tactical", "Operations"],
  ["Navigation", "Tactical", "Operations", "Engineer"],
  ["Sensors", "Navigation", "Tactical", "Operations", "Engineer"],
  ["Command", "Sensors", "Navigation", "Tactical", "Operations", "Engineer"],
  [
    "Command",
    "Sensors",
    "Navigation",
    "Tactical",
    "Operations",
    "Engineer",
    "Communications",
  ],
  [
    "Command",
    "Sensors",
    "Navigation",
    "Tactical",
    "Operations",
    "Engineer",
    "Communications",
    "Science",
  ],
  [
    "Command",
    "Sensors",
    "Navigation",
    "Tactical",
    "Operations",
    "Engineer",
    "Communications",
    "Science",
    "Security",
  ],
  [
    "Command",
    "Sensors",
    "Navigation",
    "Tactical",
    "Operations",
    "Engineer",
    "Communications",
    "Science",
    "Security",
    "Maintenance",
  ],
  [
    "Command",
    "Sensors",
    "Navigation",
    "Tactical",
    "Operations",
    "Engineer",
    "Communications",
    "Science",
    "Security",
    "Maintenance",
    "Medical",
  ],
];
export function createStationComplement({
  crewCount = 6,
  crewCaptain = true,
  ship,
  flightDirector = true,
}: {
  crewCount?: number;
  crewCaptain?: boolean;
  ship: Entity;
  flightDirector?: boolean;
}) {
  // No station set; lets create a new one.
  let stationCount = crewCount - (crewCaptain ? 1 : 0);
  // TODO: Finish this up.
  // This should take the following into account:
  //  - Whether the ship is configured to use the marauder's map
  //  - What ship systems are assigned to the ship
  const stationComplement = new StationComplementComponent({
    stations: [
      flightDirector &&
        new Station({
          name: "Flight Director",
          cards: [
            new Card({
              name: "Core",
              component: "Core",
            }),
          ],
        }),
      new Station({
        name: "Viewscreen",
        cards: [
          new Card({
            name: "Viewscreen",
            component: "Viewscreen",
          }),
        ],
      }),
      new Station({
        name: "Development",
        cards: [
          new Card({
            name: "Pilot",
            component: "Pilot",
          }),
        ],
      }),
    ].filter(function removeStation(
      station: boolean | Station
    ): station is Station {
      return Boolean(station);
    }),
  });
  return stationComplement;
}
