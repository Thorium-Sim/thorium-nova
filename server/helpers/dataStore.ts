/* istanbul ignore file */
import fsCallback from "fs";
import path from "path";
import throttle from "lodash.throttle";

const fs = fsCallback.promises;

interface IStoreOptions {
  path?: string;
  indent?: number;
  class?: any;
  initialData?: any;
  throttle?: number;
}

// A helper to make sure we don't parse any janky JSON
// It works, which is why there are so many "any"s
function json(data: any, replacer: any = null, space: number) {
  function stringify(obj: any, replacer: any, spaces: number) {
    return JSON.stringify(obj, serializer(replacer), spaces);
  }

  function serializer(replacer: any) {
    var stack: any = [],
      keys: any = [];

    const cycleReplacer = function (key: string, value: any) {
      if (stack[0] === value) return "[Circular ~]";
      return (
        "[Circular ~." + keys.slice(0, stack.indexOf(value)).join(".") + "]"
      );
    };

    return function (this: any, key: any, value: any) {
      if (stack.length > 0) {
        var thisPos = stack.indexOf(this);
        ~thisPos ? stack.splice(thisPos + 1) : stack.push(this);
        ~thisPos ? keys.splice(thisPos, Infinity, key) : keys.push(key);
        if (~stack.indexOf(value)) value = cycleReplacer.call(this, key, value);
      } else stack.push(value);

      return replacer == null ? value : replacer.call(this, key, value);
    };
  }
  return stringify(data, replacer, space);
}

function isClass(v: any) {
  return typeof v === "function" && /^\s*class\s+/.test(v.toString());
}

let isProxy = Symbol("isProxy");

interface StoreObject {
  writeFile: (force?: boolean) => Promise<void>;
  removeFile: () => Promise<void>;
  serialize?: Function;
}
export default function getStore<G extends object>(options?: IStoreOptions) {
  const {
    path: filePath,
    class: classConstructor,
    indent = 2,
    throttle: throttleDuration = 1000 * 30,
    initialData,
  } = options || {};

  // Load the data
  let _data;
  try {
    _data =
      process.env.NODE_ENV !== "test" && filePath
        ? JSON.parse(fsCallback.readFileSync(filePath, "utf8"))
        : initialData;
  } catch (err) {
    if (err.code === "EACCES") {
      err.message +=
        "\ndata-store does not have permission to load this file\n";
      throw err;
    }
  }

  if (!_data) {
    _data = initialData;
  }

  // Instantiate the object if it is a class
  // or just make a new object with the data inside
  let dataObject!: G & StoreObject;
  if (_data.length) {
    if (isClass(classConstructor)) {
      dataObject = _data.map((d: any) => new classConstructor(d));
    } else {
      dataObject = _data;
    }
    dataObject.writeFile = writeFile;
    dataObject.removeFile = removeFile;
  } else {
    if (isClass(classConstructor)) {
      dataObject = new classConstructor(_data);
      dataObject.writeFile = writeFile;
      dataObject.removeFile = removeFile;
    } else {
      dataObject = {..._data, writeFile, removeFile};
    }
  }

  async function writeFile(force = false) {
    if (process.env.NODE_ENV !== "production" && force === false) return;
    if (!filePath) {
      return;
    }
    await fs.mkdir(path.dirname(filePath), {recursive: true});
    const jsonData = json(
      dataObject.serialize ? dataObject.serialize() : dataObject,
      null,
      indent
    );
    await fs.writeFile(filePath, jsonData, {mode: 0o0600});
  }

  async function removeFile() {
    if (!filePath) return;
    const removePath = path.resolve("../", filePath);
    try {
      await fs.rmdir(removePath, {recursive: true});
    } catch (err) {
      console.error("Error removing file: ", filePath, err);
    }
  }

  const writeThrottle =
    process.env.NODE_ENV === "test"
      ? writeFile
      : throttle(writeFile, throttleDuration, {
          trailing: true,
        });

  const handler: ProxyHandler<any> = {
    get(target, key) {
      if (key === isProxy) return true;
      if (
        !target[isProxy] &&
        Object.getOwnPropertyDescriptor(target, key) &&
        typeof target[key] === "object" &&
        target[key] !== null &&
        !(target[key] instanceof Date)
      ) {
        return new Proxy(target[key], handler);
      } else {
        return target[key];
      }
    },
    set(target, name, value) {
      target[name] = value;
      writeThrottle();
      return true;
    },
    deleteProperty(target, prop) {
      if (prop in target) {
        delete target[prop];
        writeThrottle();
        return true;
      }
      return false;
    },
  };
  return new Proxy(dataObject, handler) as G & StoreObject;
}
