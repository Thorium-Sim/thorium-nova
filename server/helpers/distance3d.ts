export function distance3d(
  coord2: {x: number; y: number; z: number},
  coord1: {x: number; y: number; z: number}
) {
  const {x: x1, y: y1, z: z1} = coord1;
  let {x: x2, y: y2, z: z2} = coord2;
  return Math.sqrt((x2 -= x1) * x2 + (y2 -= y1) * y2 + (z2 -= z1) * z2);
}
