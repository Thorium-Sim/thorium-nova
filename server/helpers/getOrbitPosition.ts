import {Vector3} from "three";
export const DEG_TO_RAD = Math.PI / 180;

const axis = new Vector3(0, 0, 1);

export function getOrbitPosition({
  radius,
  eccentricity,
  orbitalArc,
  orbitalInclination,
  origin = new Vector3(),
}: OrbitPositionProps) {
  const radiusY = radius - radius * eccentricity;
  const X = radius * Math.cos(DEG_TO_RAD * orbitalArc);
  const Z = radiusY * Math.sin(DEG_TO_RAD * orbitalArc);
  const vec = new Vector3(X, 0, Z);
  const angle = orbitalInclination * DEG_TO_RAD;

  vec.applyAxisAngle(axis, angle).add(origin);
  return vec;
}

interface OrbitPositionProps {
  radius: number;
  eccentricity: number;
  orbitalArc: number;
  orbitalInclination: number;
  origin?: Vector3;
}
