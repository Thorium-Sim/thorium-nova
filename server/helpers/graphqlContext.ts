export interface GraphQLContext {
  clientId: string;
}
