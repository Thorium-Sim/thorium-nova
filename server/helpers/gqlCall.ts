import {graphql, GraphQLSchema} from "graphql";
import {Maybe} from "type-graphql";

import createSchema from "./buildSchema";
import {GraphQLContext} from "./graphqlContext";

interface Options {
  query: string;
  variables?: Maybe<{
    [key: string]: any;
  }>;
  context?: GraphQLContext;
}

let schema: GraphQLSchema;

export const gqlCall = async ({
  query,
  variables,
  context = {clientId: "test-client"},
}: Options) => {
  if (!schema) {
    schema = await createSchema();
  }
  return graphql({
    schema,
    source: query,
    contextValue: context,
    variableValues: variables,
  });
};
