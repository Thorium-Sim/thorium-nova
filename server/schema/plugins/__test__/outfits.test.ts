import {gqlCall} from "../../../helpers/gqlCall";
import path from "path";
// @ts-ignore
import {Upload} from "graphql-upload";

const fs = jest.genMockFromModule("fs") as any;

describe("Ship Plugin", () => {
  it("should query and get no results", async () => {
    const plugin = await gqlCall({
      query: `mutation CreatePlugin($name:String!) {
      pluginCreate(name:$name) {
        id
        name
      }
    }`,
      variables: {name: "My Outfit Plugin 1"},
    });
    const pluginId = plugin.data?.pluginCreate.id;
    const outfits = await gqlCall({
      query: `query Outfits($pluginId:ID!) {
      pluginOutfits(pluginId:$pluginId) {
        id
        identity {
          name
        }
      }
    }`,
      variables: {pluginId},
    });
    expect(outfits.data?.pluginOutfits).toEqual([]);
  });
  it("should create a new outfit", async () => {
    const plugin = await gqlCall({
      query: `mutation CreatePlugin($name:String!) {
      pluginCreate(name:$name) {
        id
        name
      }
    }`,
      variables: {name: "My Outfit Plugin 2"},
    });
    const pluginId = plugin.data?.pluginCreate.id;
    const newOutfit = await gqlCall({
      query: `mutation AddOutfit($pluginId:ID!) {
        pluginAddOutfit(pluginId:$pluginId,  ability:thrusters) {
          id
        }
      }`,
      variables: {pluginId},
    });
    const outfitId = newOutfit.data?.pluginAddOutfit.id;
    const outfits = await gqlCall({
      query: `query Outfits($pluginId:ID!) {
      pluginOutfits(pluginId:$pluginId) {
        id
        identity {
          name
        }
        thrusters {
          thrusting
        }
       
      }
    }`,
      variables: {pluginId},
    });
    expect(outfits.data?.pluginOutfits.length).toEqual(1);
  });
});
