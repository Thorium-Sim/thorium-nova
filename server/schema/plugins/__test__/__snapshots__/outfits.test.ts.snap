// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`outfit Plugin getOutfitComponents should return the correct components for each outfit 1`] = `
Array [
  Object {
    "component": [Function],
    "defaultValue": Object {
      "outfitType": "warpEngines",
    },
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
]
`;

exports[`outfit Plugin getOutfitComponents should return the correct components for each outfit 2`] = `
Array [
  Object {
    "component": [Function],
    "defaultValue": Object {
      "outfitType": "impulseEngines",
    },
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
]
`;

exports[`outfit Plugin getOutfitComponents should return the correct components for each outfit 3`] = `
Array [
  Object {
    "component": [Function],
    "defaultValue": Object {
      "outfitType": "thrusters",
    },
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
]
`;

exports[`outfit Plugin getOutfitComponents should return the correct components for each outfit 4`] = `
Array [
  Object {
    "component": [Function],
    "defaultValue": Object {
      "outfitType": "navigation",
    },
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
]
`;

exports[`outfit Plugin getOutfitComponents should return the correct components for each outfit 5`] = `
Array [
  Object {
    "component": [Function],
    "defaultValue": Object {
      "outfitType": "jumpDrive",
    },
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
]
`;

exports[`outfit Plugin getOutfitComponents should return the correct components for each outfit 6`] = `
Array [
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
  Object {
    "component": [Function],
  },
]
`;
